From f1153100884e580dce331e8dd56b260ffb2f04d5 Mon Sep 17 00:00:00 2001
From: Emerson Knapp <eknapp@amazon.com>
Date: Tue, 2 Feb 2021 17:47:46 -0800
Subject: [PATCH] Don't install internal headers

Signed-off-by: Emerson Knapp <eknapp@amazon.com>
---
 build/cmake/lib/CMakeLists.txt | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/build/cmake/lib/CMakeLists.txt b/build/cmake/lib/CMakeLists.txt
index 7adca875..0c2d777e 100644
--- a/build/cmake/lib/CMakeLists.txt
+++ b/build/cmake/lib/CMakeLists.txt
@@ -147,10 +147,6 @@ endif ()
 # install target
 install(FILES
     ${LIBRARY_DIR}/zstd.h
-    ${LIBRARY_DIR}/deprecated/zbuff.h
-    ${LIBRARY_DIR}/dictBuilder/zdict.h
-    ${LIBRARY_DIR}/dictBuilder/cover.h
-    ${LIBRARY_DIR}/common/zstd_errors.h
     DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
 
 if (ZSTD_BUILD_SHARED)
-- 
2.17.1

